src/VecSim/spaces/spaces.cpp

```diff
@@ -4,36 +4,77 @@
  *the Server Side Public License v1 (SSPLv1).
  */
 
+#include "VecSim/types/bfloat16.h"
 #include "VecSim/spaces/space_includes.h"
 #include "VecSim/spaces/spaces.h"
 #include "VecSim/spaces/IP_space.h"
 #include "VecSim/spaces/L2_space.h"
-namespace spaces {
-
-void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<float> *out_func,
-                 unsigned char *alignment) {
-
-    if (metric == VecSimMetric_Cosine || metric == VecSimMetric_IP) {
+#include "VecSim/spaces/normalize/normalize_naive.h"
 
-        *out_func = IP_FP32_GetDistFunc(dim, nullptr, alignment);
+#include <stdexcept>
+namespace spaces {
 
-    } else if (metric == VecSimMetric_L2) {
+// Set the distance function for a given data type, metric and dimension. The alignment hint is
+// determined according to the chosen implementation and available optimizations.
 
-        *out_func = L2_FP32_GetDistFunc(dim, nullptr, alignment);
+template <>
+dist_func_t<float> GetDistFunc<vecsim_types::bfloat16, float>(VecSimMetric metric, size_t dim,
+                                                              unsigned char *alignment) {
+    switch (metric) {
+    case VecSimMetric_Cosine:
+    case VecSimMetric_IP:
+        return IP_BF16_GetDistFunc(dim, nullptr, alignment);
+    case VecSimMetric_L2:
+        return L2_BF16_GetDistFunc(dim, nullptr, alignment);
+    default:
+        throw std::invalid_argument("Invalid metric");
     }
 }
 
-void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<double> *out_func,
-                 unsigned char *alignment) {
+template <>
+dist_func_t<float> GetDistFunc<float, float>(VecSimMetric metric, size_t dim,
+                                             unsigned char *alignment) {
+    switch (metric) {
+    case VecSimMetric_Cosine:
+    case VecSimMetric_IP:
+        return IP_FP32_GetDistFunc(dim, nullptr, alignment);
+    case VecSimMetric_L2:
+        return L2_FP32_GetDistFunc(dim, nullptr, alignment);
+    default:
+        throw std::invalid_argument("Invalid metric");
+    }
+}
 
-    if (metric == VecSimMetric_Cosine || metric == VecSimMetric_IP) {
+template <>
+dist_func_t<double> GetDistFunc<double, double>(VecSimMetric metric, size_t dim,
+                                                unsigned char *alignment) {
+    switch (metric) {
+    case VecSimMetric_Cosine:
+    case VecSimMetric_IP:
+        return IP_FP64_GetDistFunc(dim, nullptr, alignment);
+    case VecSimMetric_L2:
+        return L2_FP64_GetDistFunc(dim, nullptr, alignment);
+    default:
```

Comment:
Why add a default here? Don't we compile with "warn incomplete switch enum"?

Reply:
The compiler will complain that the function doesn't return a value

Reply:
so consider moving the exception outside the switch, so future metrics won't automatically get covered by the default in the future

Reply:
right good point